Document OS-specific approaches for noarch packages (#1839)
conda-forge/conda-forge.github.io: src/maintainer/knowledge_base.rst

  which builds on Linux `and` Windows, with ``build_number`` offsets to create a pair packages, like
  ``dataclasses``.

.. hint::
  
  You can build platform-specific ``noarch`` packages to include runtime requirements depending on the target OS.
  See mini-tutorial below.

If an existing python package qualifies to be converted to a noarch package, you can request the required changes
by opening a new issue and including ``@conda-forge-admin, please add noarch: python``.

.. _os_specific_noarch:

Noarch packages with OS-specific dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

It is possible to build ``noarch`` packages with runtime requirements that depend on the target OS (Linux, Windows,
MacOS), regardless the architecture (amd64, ARM, PowerPC, etc). This approach relies on four concepts:

> 1.  Virtual packages. Prefixed with a double underscore, they are used by conda to represent properties of the running system 

Reviewer: ```suggestion
1.  Virtual packages. Prefixed with a double underscore, they are used by conda
    when running the dependency resolver to represent properties of the currently
    activated environment 
```

Author: Same as below, technically they are environment independent unless overridden.

    as constraints for the solver. We will use ``__linux``, ``__win`` or ``__osx``, which are only present when

>     the running platform is Linux, Windows, or MacOS, respectively. ``__unix`` is present in both Linux and MacOS. Note

Reviewer: ```suggestion
    the activated environment is Linux, Windows, or MacOS, respectively. ``__unix`` is present in both Linux and MacOS. Note
```

Author: I think one can only change the virtual packages with `CONDA_OVERRIDE_*` env variables. Otherwise, regardless the active environment, it will still report the native platform type.

Reviewer: OK, thanks for the clarification. I don't know if "running platform" is defined somewhere?

Reviewer: also, it is not clear to me from "as constraints for the solver" that this happens at runtime, and not while building the package.

Author: I'll rephrase, thanks!

would never be true). Fortunately, we can change the default behaviour in ``conda-forge.yml``:

.. code-block:: yaml

  # conda-forge.yml
  noarch_platforms:
    - linux-64
    - win-64

This will provide two runners per package! But since we are using selectors in ``conda_build_config.yaml``,
only one is true at a time. Perfect! All these changes require a feedstock rerender to be applied:
:ref:`_dev_update_rerender`.

Last but not least, what if you need conditional dependencies on all three operating systems? Do it like this:

.. code-block:: yaml

  # recipe/meta.yaml
  name: package
  source:
   ...
  build:
    number: 0
    noarch: python
  requirements:
    ...
    run:
      - python
      - numpy
      - __{{ target_os }}
      {% if target_os == 'osx' %}
      - osx-only-dependency
      {% elif target_os == 'win' %}
      - windows-only-dependency
      {% else %}
      - linux-only-dependency
      {% endif %}

.. code-block:: yaml

  # recipe/conda_build_config.yaml
  target_os:
    - linux  # [linux]
    - osx    # [osx]
    - win    # [win]

.. code-block:: yaml

  # conda-forge.yml
  noarch_platforms:
    - linux-64
    - osx-64
    - win-64

> Again, d

Reviewer: ```suggestion

```

  See mini-tutorial below.

If an existing python package qualifies to be converted to a noarch package, you can request the required changes
by opening a new issue and including ``@conda-forge-admin, please add noarch: python``.

.. _os_specific_noarch:

Noarch packages with OS-specific dependencies
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

It is possible to build ``noarch`` packages with runtime requirements that depend on the target OS
(Linux, Windows, MacOS), regardless the architecture (amd64, ARM, PowerPC, etc). This approach
relies on three concepts:

1.  `Virtual packages <https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html>`__. 
    Prefixed with a double underscore, they are used by conda to represent system properties as
    constraints for the solver at install-time. We will use ``__linux``, ``__win`` or ``__osx``,
    which are only present when the running platform is Linux, Windows, or MacOS, respectively.
    ``__unix`` is present in both Linux and MacOS. Note that this feature is **only fully available
    on conda 4.10 or above**.

> 2.  ``conda-forge.ymls``'s :ref:`noarch_platforms` option.

Reviewer 2: ```suggestion
2.  ``conda-forge.yml``'s :ref:`noarch_platforms` option.
```

Author: Good catch, thanks!

The idea is to generate different noarch packages for each OS needing different dependencies.
Let's say you have a pure Python package, perfectly eligible for ``noarch: python``, but on Windows
it requires ``windows-only-dependency``. You might have something like:

.. code-block:: yaml
  :caption: recipe/meta.yaml (original)

  name: package
  source:
    # ...
  build:
    number: 0
  requirements:
    # ...
    run:
      - python
      - numpy
      - windows-only-dependency  # [win]

Being non-noarch, this means that the build matrix will include at least 12 outputs: three platforms,
times four Python versions. This gets worse with arm64, aarch64 and ppc64le in the mix. We can get it down
to two outputs if replace it with this other approach!

.. code-block:: yaml+jinja
  :caption: recipe/meta.yaml (modified)

  name: package
  source:
    # ...
  build:
    number: 0
    # You NEED to include the platform in the build string to avoid hash collisions
    string: "unix_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"  # [unix]

>     string: "win_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"   # [win]

Reviewer 3: Would it be possible to avoid manipulating the `string` and instead rely on different hashes (note Jaime is mentioning something similar elsewhere)?

The main issue with `string` is pretty finicky. Users need to be very careful to replicate conda-build's behavior and if they missing something they could wind up with duplicate packages (which sometimes can both be accepted by Anaconda.org with neither being downloadable).

Author: This will be needed until https://github.com/conda/conda-build/pull/4606 lands and a new conda-build release is cut.

Is it ok to merge as is or would you rather have a little admonition mentioning this caveat?

Reviewer 3: My understanding from the thread above is using `conda_build_config.yaml` meant we would not need this.

Fixing conda-build to not need either is certainly better. That said, between setting a `string` and using `conda_build_config.yaml`, would prefer the latter (fewer foot guns).

Author: I am partial to `conda_build_config.yaml` too as we transition to the new conda-build version.

    noarch: python
  requirements:
    # ...
    run:
      - python
      - numpy
      - __unix  # [unix]
      - __win   # [win]
      - windows-only-dependency  # [win]

Do not forget to specify the platform virtual packages with their selectors!
Otherwise, the solver will not be able to choose the variants correctly.

By default, conda-forge will only build ``noarch`` packages on a ``linux_64`` CI runner, so
only the ``# [unix]`` selectors would be true. However, we can change this behaviour using
the ``noarch_platforms`` option in ``conda-forge.yml``:

.. code-block:: yaml
  :caption: conda-forge.yml

  noarch_platforms:
    - linux_64
    - win_64

This will provide two runners per package! Perfect! All these changes require a
feedstock rerender to be applied. See :ref:`dev_update_rerender`.

If you need conditional dependencies on all three operating systems, this is how you do it:

.. code-block:: yaml+jinja
  :caption: recipe/meta.yaml

  name: package
  source:
    # ...
  build:
    number: 0
    # You NEED to include the platform in the build string to avoid hash collisions

>     string: "{{ SUBDIR.split('-')[0] }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"

Reviewer 4: I'm getting a linter error with this: https://github.com/conda-forge/poetry-feedstock/pull/80#issuecomment-1354686510
Commit: https://github.com/conda-forge/poetry-feedstock/pull/80/commits/e2abf93666388971b43d25581a84b3356a2c8d41

Author: Hm, we might need to run `conda smithy recipe-lint` locally to see the actual error.

Reviewer 4: ```python
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/mambaforge/base/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 670, in main
    args.subcommand_func(args)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 514, in __call__
    lints, hints = lint_recipe.main(
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/lint_recipe.py", line 978, in main
    content = render_meta_yaml("".join(fh))
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/utils.py", line 90, in render_meta_yaml
    content = env.from_string(text).render(context)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 16, in top-level template code
TypeError: 'str' object is not callable
```
This is the full error.

Author: Ah, I think we need to split `SUBDIR` in a different `set` expression. 
See https://stackoverflow.com/questions/30515456/split-string-into-list-in-jinja

So we should have:

```yaml
{% set os_prefix, arch_suffix = SUBDIR.split('-', 1) %}
string: "{{ os_prefix }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Reviewer 4: Hmm, still getting the same error:

```python
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/mambaforge/base/bin/conda-smithy", line 10, in <module>
    sys.exit(main())
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 670, in main
    args.subcommand_func(args)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/cli.py", line 514, in __call__
    lints, hints = lint_recipe.main(
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/lint_recipe.py", line 978, in main
    content = render_meta_yaml("".join(fh))
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/conda_smithy/utils.py", line 90, in render_meta_yaml
    content = env.from_string(text).render(context)
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
    self.environment.handle_exception()
  File "/opt/homebrew/Caskroom/mambaforge/base/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 3, in top-level template code
TypeError: 'str' object is not callable
```

Recipe:

```yaml
{% set name = "poetry" %}
{% set version = "1.3.1" %}
{% set os_prefix, arch_suffix = SUBDIR.split('-', 1) %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/poetry-{{ version }}.tar.gz
  sha256: fde98462ad5dc9879393157da93092206a3411117e25a4761a41c6d08f31aea8

build:
  noarch: python
  number: 1
  # You NEED to include the platform in the build string to avoid hash collisions
  string: "{{ os_prefix }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
  script: {{ PYTHON }} -m pip install . -vv
  entry_points:
    - poetry = poetry.console.application:main
```

Author: Ugh, turns out `SUBDIR` is not defined. But `target_platform` works:

```yaml
  string: "{{ target_platform.split('-')[0] }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
```

Still, a bit ugly...

Author: Changed the PR so it's explicit about the platforms and doesn't use that much Jinja in the string.

2.  ``conda-forge.yml``'s :ref:`noarch_platforms` option.

The idea is to generate different noarch packages for each OS needing different dependencies.
Let's say you have a pure Python package, perfectly eligible for ``noarch: python``, but on Windows
it requires ``windows-only-dependency``. You might have something like:

.. code-block:: yaml
  :caption: recipe/meta.yaml (original)

  name: package
  source:
    # ...
  build:
    number: 0
  requirements:
    # ...
    run:
      - python
      - numpy
      - windows-only-dependency  # [win]

Being non-noarch, this means that the build matrix will include at least 12 outputs: three platforms,
times four Python versions. This gets worse with arm64, aarch64 and ppc64le in the mix. We can get it down
to two outputs if replace it with this other approach!

.. code-block:: yaml+jinja
  :caption: recipe/meta.yaml (modified)

  name: package
  source:
    # ...
  build:
    number: 0
    # You NEED to include the platform in the build string to avoid hash collisions
    string: "unix_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"  # [unix]
    string: "win_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"   # [win]
    noarch: python
  requirements:

>     # ...
>     run:
>       - python
>       - numpy

Reviewer 5: ```suggestion
    host:
      - python >=3.7
      # ...
    run:
      - python >=3.7
      - numpy
```

don't we want to encourage lower bounds on noarch packages?

Let's say you have a pure Python package, perfectly eligible for ``noarch: python``, but on Windows
it requires ``windows-only-dependency``. You might have something like:

.. code-block:: yaml
  :caption: recipe/meta.yaml (original)

  name: package
  source:
    # ...
  build:
    number: 0
  requirements:
    # ...
    run:
      - python
      - numpy
      - windows-only-dependency  # [win]

Being non-noarch, this means that the build matrix will include at least 12 outputs: three platforms,
times four Python versions. This gets worse with arm64, aarch64 and ppc64le in the mix. We can get it down
to two outputs if replace it with this other approach!

.. code-block:: yaml+jinja
  :caption: recipe/meta.yaml (modified)

  name: package
  source:
    # ...
  build:
    number: 0
    # You can include target_os in the build string for easier identification of builds
    string: "{{ target_os }}_pyh{{ PKG_HASH }}_{{ PKG_BUILDNUM }}"
    noarch: python
  requirements:
    # ...
    run:
      - python
      - numpy
      - __{{ target_os }}
      {% if target_os == 'win' %}
      - windows-only-dependency

>       {% endif %}

Author: Ok my main concern about not using `conda_build_config.yaml`, @isuruf: 

* The package hashes will be the same (see https://github.com/conda-forge/click-feedstock/pull/44), so the build string NEEDS to be prefixed with the platform value. If the user forgets about this, it will result in a filename collision. This should be checked by the linter, at the very least. I am unaware if sharing a hash is problematic for other cases.

>       {% endif %}

Reviewer 3: I share this concern and would prefer not using `string`. Explained in more detail below ( https://github.com/conda-forge/conda-forge.github.io/pull/1839#discussion_r1008679066 )

>       {% endif %}

Reviewer 6: See https://github.com/conda/conda-build/pull/4606

Cool! Where does ``target_os`` come from? We need to define it in ``conda_build_config.yaml``.
Note how the values have been chosen carefully so they match the virtual packages names:

.. code-block:: yaml
  :caption: recipe/conda_build_config.yaml

  target_os:
    - unix  # [unix]
    - win   # [win]

By default, conda-forge will only build ``noarch`` packages on a ``linux-64`` CI runner, so
the ``target_os`` matrix would only provide the ``unix`` value (because the ``# [win]`` selector

> would never be true). Fortunately, we can change the default behaviour in ``conda-forge.yml``:

Reviewer: ```suggestion
would never be true). Fortunately, we can change the default behaviour in ``./conda-forge.yml``:
```
